crates/extensions/c8y_mapper_ext/src/operations/handler.rs

```diff
@@ -138,23 +141,70 @@ impl OperationHandler {
+        let topic = Arc::from(message.message.topic.name.as_str());
 
-        let running_operation = self.running_operations.remove(&topic);
+        let status = match GenericCommandState::from_command_message(&message.message) {
+            Ok(command) if command.status.is_empty() && is_clearing_message(&message.message) => {
+                None
+            }
+            Ok(command) => Some(command.status),
+            Err(err) => {
+                error!(%err, ?message, "could not parse command payload");
+                return;
+            }
+        };
 
-        let running_operation =
-            running_operation.unwrap_or_else(|| RunningOperation::spawn(Arc::clone(&self.context)));
+        let current_operation = self.running_operations.entry(topic);
 
-        let operation_status = running_operation
-            .update(message)
-            .await
-            .expect("operation task should not panic");
+        match current_operation {
+            Entry::Vacant(entry) => {
+                let Some(status) = status else {
+                    debug!("unexpected clearing message");
+                    return;
+                };
+
+                let context = Arc::clone(&self.context);
+                let handle = tokio::spawn(async move { context.update(message).await });
 
-        match operation_status {
-            OperationStatus::Ongoing(operation) => {
-                self.running_operations.insert(topic, operation);
+                let running_operation = RunningOperation { handle, status };
+
+                entry.insert(running_operation);
+            }
+
+            Entry::Occupied(entry) => {
+                if status.as_ref().is_some_and(|s| *s == entry.get().status) {
+                    debug!(
+                        "already handling operation message with this topic and status, ignoring"
+                    );
+                    return;
+                }
+
+                let (key, operation) = entry.remove_entry();
+
+                // TODO(marcel):
+                //- what happens when clearing message is posted by other client while we're
+                // handling a message? E.g. if we're doing a download to handle SUCCESSFUL status, do
+                // we abort or keep going?
+                // - what if we get a message that moves back status, i.e. SUCCESSFUL -> EXECUTING?
+
+                // if handling clearing message, just join task and remove
+                let Some(status) = status else {
+                    operation
+                        .handle
+                        .await
+                        .expect("operation task should not panic");
+                    return;
+                };
+
+                // we got a new status, await we handled the previous one and handle the new one
+                let context = Arc::clone(&self.context);
+                let handle = tokio::spawn(async move {
```

Comment:
I didn't funny understand why you moved to this model where new tasks are spawned for each state transition as opposed to the previous model of single task per operation.

Reply:
Because we need to extract status from every message to prevent illegal state transitions, it makes sense to do it once in the handle method of the handler for all operation types. And for that we need to know what is the current state of the operation.
It also achieves separation of concerns: operation handler keeps state and handles the lifecycle of the operation, and `context.update()` is just a function that sends a smartrest response when given an operation message.
Ultimately, we only spawn a task so we that we can eventually generate a smartrest response (possibly after downloads and uploads) for that operation message. The logic to determine if a state transition is valid should be lifted up.